de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/SarosComponent.java

```diff
@@ -47,7 +71,54 @@ public SarosComponent(final Project project) {
       LogLog.error("could not load saros property file 'saros.properties'", e);
     }
 
-    IntellijProjectLifecycle.getInstance(project).start();
+    intellijProjectLifecycle.start();
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Does nothing for the Saros plugin.
+   */
+  @Override
+  public void disposeComponent() {
+    // NOP
+  }
+
+  @Override
+  @NotNull
+  public String getComponentName() {
+    return "Saros";
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Does nothing for the Saros plugin as the initialization is already done in {@link
+   * #initComponent()}.
+   */
+  @Override
+  public void projectOpened() {
```

Comment:
Isnt this called multiple times. As far as I can remember you should not expect that projectClosed() is called before projectOpened() is called.

To keep the issue simple: Isn't there a option to shutdown Saros before a new project is either opened or the user switches the project ?

Reply:
@srossbach Yes, this gets called multiple times and yes, this can get called before the last project was opened.

The problem is that each setup method is called for ever project that is opened (CTOR, `initComponent()`, and `projectOpened()`), even if there is already an open project.

I was not sure how to handle the case that a user opens a second project, so my previous inclination was to exclude the case through the release notes (which is not a very good handling of the case). But not mentioning the problem at all was a mistake. My bad.

But thinking about it now, my preferred handling of this case would be that Saros only acts on the project that was opened first. Any additional projects will be ignored as long as there is currently an open project. (Later on, we can deal with this by always using the project that is focused by the user when trying to start a session, but I don't think this should be implemented for the first alpha. This should be done when moving to Intellij application scope.)
This would mean that we would have to check whether there is currently already a project object in the context and only add the opened project if there isn't one (no possible with the current handling, but easily adjustable; would also mean null-checks, but violations here would have thrown disposed access assertions instead anyway).

This still leaves us with a couple of problems that I don't know how to handle gracefully:
- What do we do when the user closes the shareable project? This gets worse when there are multiple open projects. Even if we store all project object, which one do we choose?
- The [`NegotiationHandler`](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/eventhandler/NegotiationHandler.java#L111) still uses a deprecated method that chooses the project object by looking at the current focus. I am not sure if it would be possible to create a weird corner case with this that leads to problems.

Furthermore, we would have to adjust [`SarosFileShareGroup`](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/menu/SarosFileShareGroup.java#L84) to check whether the focused project is shareable (or update the context).

In general, there is the potential for a lot of corner cases when allowing multiple open project and I don't think we can handle them well with the current setup. So I am not sure whether allowing multiple open projects is really a good idea before moving the plugin to application scope. (I know this somewhat contradicts my statement above, but I am pondering as I am writing :wink:).

We can still release a second version not long after , but I would like to get something out there soon-ish so that we at least appear on the radar and to show that something is actually happening.

Reply:
I also think that the restriction of "one open project at a time" is acceptable for the first alpha version. This (change of plugin licecycle to application lifecycle) is also a point that should we document as "ongoing work".